### PR TITLE
supporting distributed training on 1, 2, 4, and 8 machines and supporting fp16

### PR DESCRIPTION
**Pull request by sandyhouse**
supporting distributed training on 1, 2, 4, and 8 machines and supporting fp16

### CODE REVIEW

**Review by ForFishes — APPROVED** [review]

File: PaddleCV/image_classification/fast_imagenet/train.py [hunk 1]
```diff
@@ -12,104 +12,93 @@
...
 import os
 import traceback
+os.environ['FLAGS_sync_nccl_allreduce'] = '1'
 
 import numpy as np
 import math
 import reader
 import paddle
 import paddle.fluid as fluid
 import paddle.fluid.profiler as profiler
```
Comment (sandyhouse) [inline PaddleCV/image_classification/fast_imagenet/train.py:30]: Now, paddle.distributed.launch sets this environment variable to '0'. So this statement is needed if you want to enable this feature.

File: PaddleCV/image_classification/fast_imagenet/train.py [hunk 2]
```diff
@@ -12,104 +12,93 @@
...
+    parser = argparse.ArgumentParser(description="Training ImageNet in Tens of Minutes.")
+    add_arg = functools.partial(add_arguments, argparser=parser)
+    add_arg('total_images',     int,   1281167,     "Number of training images.")
+    add_arg('num_epochs',       int,   120,         "Maximum number of epochs to run.")
+    add_arg('model_save_dir',   str,   "output",    "Directory to save models")
+    add_arg('data_dir',         str,   "dataset/",  "Directory for dataset.")
+    add_arg('fp16',             bool,  False,       "Whether to use half precision training.")
+    add_arg('scale_loss',       float, 64.0,        "Scale loss for fp16.")
+    add_arg('start_test_pass',  int,   0,           "At which pass to start test.")
+    add_arg('use_hallreduce',   bool,  False,       "Whether to use hierarchical allreduce.")
```
Comment (sandyhouse) [inline PaddleCV/image_classification/fast_imagenet/train.py:53]: These variables are used. use_hallreduce is used in dist_utils.py

**Review by ForFishes — APPROVED** [review]

**Review by ForFishes — APPROVED** [review]
LGTM

**Review by ForFishes — APPROVED** [review]

**Review by ForFishes — APPROVED** [review]

File: PaddleCV/image_classification/fast_imagenet/dist_utils.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+#
+#Unless required by applicable law or agreed to in writing, software
+#distributed under the License is distributed on an "AS IS" BASIS,
+#WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#See the License for the specific language governing permissions and
+#limitations under the License.
+
+import os
+import paddle.fluid as fluid
+
```
Comment (ForFishes) [inline PaddleCV/image_classification/fast_imagenet/dist_utils.py]: test

**Review by ForFishes — APPROVED** [review]

**Review by ForFishes — APPROVED** [review]

**Pull request closed without merging**